cert/cert.go

```diff
@@ -65,6 +67,39 @@ func NewServer(caCertPEM, caKeyPEM string, expiry time.Time, hostnames []string)
 	})
 }
 
+var hexAlphabet = [16]byte{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'}
+
+// Fingerprint returns a human-readable SHA-256 fingerprint for a certificate
+// stored in the PEM format. The returned fingerprint matches the output of:
+// openssl x509 -noout -fingerprint -sha256 -inform pem -in cert.pem
+func Fingerprint(certPEM string) (string, error) {
+	block, _ := pem.Decode([]byte(certPEM))
+	if block == nil {
+		return "", errors.New("input does not contain a valid certificate in PEM format")
+	}
+
+	if _, err := x509.ParseCertificate(block.Bytes); err != nil {
```

Comment:
What's the reason to match openssl directly, surely we can sum256 the cert?

```go
cert, err := x509.ParseCertificate(block.Bytes)
if err != nil {
    return "", errors.Annotate(err, "cannot calculate certificate fingerprint")
}
return fmt.Sprintf("%x", sha256.Sum256(cert.Raw)), nil
```

Reply:
When we redirect users to the new controller for a migrated model we will also list the fingerprint for the new controller's CA. Then, when the user tries to `juju login` to the new controller we will display the fingerprint and ask them to trust it before proceeding.

As this is used in a purely client-facing scenario it made sense for me to format the fingerprint so it's easier to compare. I mentioned openssl in the comment more as an example; this format (or the variant with space as a delimiter) is pretty common out there.